### PR TITLE
Make waiting VolumeSnapshot to ready process parallel.

### LINKED ISSUE #4879: Make waiting VolumeSnapshot to ready process parallel.
**Describe the problem/challenge you have**
[A description of the current limitation/problem/challenge that you are experiencing.]
When CSI feature is enabled, CSI plugin will use BackupItemAction to take CSI snapshot of PV. Before https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/100, CSI plugin only wait the VolumeSnapshotContent is created, before reporting snapshotting success. It's not correct, because after VSC creation, there is still possibility error causing snapshotting failure on cloud.
But the process waiting for VSC turning ready takes a lot of time, after that PR, backup completion time increased per testing on developing environment, so need to make it parallel to reduce waiting time.
It's not easy to make BackupItemAction parallel by now, so choosing to add the logic in backup controller and revert https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/100.
After BackupItemAction parallel is implemented, add the ReadyToUse checking code back into CSI plugin.

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Velero version (use `velero version`):
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

### PR DESCRIPTION
**Pull request by blackpiglet**
Signed-off-by: Xun Jiang <jxun@vmware.com>

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #4879 

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/controller/backup_controller.go [hunk 1]
```diff
@@ -865,3 +878,42 @@ func encodeToJSONGzip(data interface{}, desc string) (*bytes.Buffer, []error) {
 
 	return buf, nil
 }
+
+// Waiting for VolumeSnapshot ReadyTosue to true is time consuming. Try to make the process parallel by
+// using goroutine here instead of waiting in CSI plugin, because it's not easy to make BackupItemAction
+// parallel by now. After BackupItemAction parallel is implemented, this logic should be moved to CSI plugin
+// as https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/100
+func (c *backupController) checkVolumeSnapshotReadyToUse(volumesnapshots []*snapshotv1api.VolumeSnapshot) error {
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:886]: IMO  it's better set the context as the first param of this func so when we refactor controller the context of reconcile can be passed.

File: pkg/controller/backup_controller.go [hunk 2]
```diff
@@ -865,3 +878,42 @@ func encodeToJSONGzip(data interface{}, desc string) (*bytes.Buffer, []error) {
...
+					return false, errors.Wrapf(err, fmt.Sprintf("failed to get volumesnapshot %s/%s", volumeSnapshot.Namespace, volumeSnapshot.Name))
+				}
+				if tmpVS.Status == nil || tmpVS.Status.BoundVolumeSnapshotContentName == nil || !boolptr.IsSetToTrue(tmpVS.Status.ReadyToUse) {
+					log.Infof("Waiting for CSI driver to reconcile volumesnapshot %s/%s. Retrying in %ds", volumeSnapshot.Namespace, volumeSnapshot.Name, interval/time.Second)
+					return false, nil
+				}
+
+				return true, nil
+			})
+			if err != nil {
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:906]: you can just return err, so only leave this chunk:
```
if err == wait.ErrWaitTimeout {
	log.Errorf("Timed out awaiting reconciliation of volumesnapshot %s/%s", volumeSnapshot.Namespace, volumeSnapshot.Name)
}
return err
				```

File: pkg/controller/backup_controller.go [hunk 3]
```diff
@@ -865,3 +878,42 @@ func encodeToJSONGzip(data interface{}, desc string) (*bytes.Buffer, []error) {
...
+				if err == wait.ErrWaitTimeout {
+					log.Errorf("Timed out awaiting reconciliation of volumesnapshot %s/%s", volumeSnapshot.Namespace, volumeSnapshot.Name)
+				}
+				return err
+			}
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:915]: `return eg.Wait()`

File: pkg/controller/backup_controller.go [hunk 4]
```diff
@@ -638,6 +645,12 @@ func (c *backupController) runBackup(backup *pkgbackup.Request) error {
 				backupLog.Error(err)
 			}
 
+			err = c.checkVolumeSnapshotReadyToUse(volumeSnapshots)
+			if err != nil {
+				backupLog.Errorf("fail to wait VolumeSnapshot change to Ready: %s", err.Error())
+				return err
```
Comment (reasonerjt) [inline pkg/controller/backup_controller.go:651]: Could you double check the impact of returning err here?
I'm not very sure, but it seems we should just log the error and let it move on?

Comment (blackpiglet) [reply pkg/controller/backup_controller.go:651]: Modified. Record an error in backup log and move on with next backup process.

File: pkg/controller/backup_controller.go (hunk 2 above)
Comment (blackpiglet) [reply pkg/controller/backup_controller.go:906]: Modified.

File: pkg/controller/backup_controller.go (hunk 3 above)
Comment (blackpiglet) [reply pkg/controller/backup_controller.go:915]: Modified.

File: pkg/controller/backup_controller.go (hunk 1 above)
Comment (blackpiglet) [reply pkg/controller/backup_controller.go:886]: Modified.

**Comment by blackpiglet** [comment]
After rebase with latest main branch, the PR contains other code change. Close this and create new one.

**Pull request closed without merging**